Single block loops not placed into loop region

From `DEBUGGRAPH=1 pytest numba_rvsdg/tests/test_mock_asm.py::test_mock_scfg_fuzzer_case36` of #42 .

Single block loops are not restructured:

CFG:
![Screenshot 2023-04-18 at 2 12 39 PM](https://user-images.githubusercontent.com/1929845/232881171-6ad32ffd-abe6-46bd-beef-717f7f291a75.png)
After loop restructuring
![Screenshot 2023-04-18 at 2 13 31 PM](https://user-images.githubusercontent.com/1929845/232881329-c8585a09-5b9a-427d-81ac-e6933af74714.png)

The following patch will fix this case:

```
diff --git i/numba_rvsdg/core/transformations.py w/numba_rvsdg/core/transformations.py
index b3d59811b6..e991de2ba2 100644
--- i/numba_rvsdg/core/transformations.py
+++ w/numba_rvsdg/core/transformations.py
@@ -143,7 +143,7 @@ def loop_restructure_helper(scfg: SCFG, loop: Set[Label]):
                     # matters in this case.
                     new_jt[new_jt.index(jt)] = synth_assign
                 # If the target is the loop_head
-                elif jt in headers and label not in doms[jt]:
+                elif jt in headers: #and label not in doms[jt]:
                     # Create the assignment and record it
                     synth_assign = SynthenticAssignment(str(scfg.clg.new_index()))
                     new_blocks.add(synth_assign)
```

And yield the following, correct, result:

<img width="591" alt="Screen Shot 2023-04-20 at 15 26 45" src="https://user-images.githubusercontent.com/41563/233383645-c1778a35-e05c-4d57-828f-9177964d8258.png">


However, this breaks the fig.3  test with:

```
    def replace_jump_targets(self, jump_targets: Tuple) -> "BasicBlock":
        fallthrough = len(jump_targets) == 1
        old_branch_value_table = self.branch_value_table
        new_branch_value_table = {}
        for target in self.jump_targets:
            if target not in jump_targets:
                # ASSUMPTION: only one jump_target is being updated
                diff = set(jump_targets).difference(self.jump_targets)
>               assert len(diff) == 1
E               AssertionError

numba_rvsdg/core/datastructures/basic_block.py:94: AssertionError
```
```

--- a/numba_rvsdg/core/transformations.py
+++ b/numba_rvsdg/core/transformations.py
@@ -136,7 +136,7 @@ def loop_restructure_helper(scfg: SCFG, loop: Set[str]):
                     # matters in this case.
                     new_jt[new_jt.index(jt)] = synth_assign
                 # If the target is the loop_head
-                elif jt in headers and name not in doms[jt]:
+                elif jt in headers and (name not in doms[jt] or name == jt):
                     # Create the assignment and record it
                     synth_assign = scfg.name_gen.new_block_name(block_names.SYNTH_ASSIGN)
                     new_blocks.add(synth_assign)

--- a/numba_rvsdg/tests/test_transforms.py
+++ b/numba_rvsdg/tests/test_transforms.py
@@ -525,7 +525,7 @@ class TestLoopRestructure(SCFGComparator):
         "0":
             jt: ["1"]
         "1":
-            jt: ["2", "5"]
+            jt: ["5", "2"]
         "2":
             jt: ["6", "7"]
         "3":

--- a/numba_rvsdg/tests/test_transforms.py
+++ b/numba_rvsdg/tests/test_transforms.py
@@ -510,6 +510,41 @@ class TestLoopRestructure(SCFGComparator):
         loop_restructure_helper(original_scfg, set({block_dict["1"], block_dict["2"]}))
         self.assertSCFGEqual(expected_scfg, original_scfg)
 
+    def test_multi_back_edge_with_backedge_from_header(self):
+        original = """
+        "0":
+            jt: ["1"]
+        "1":
+            jt: ["1", "2"]
+        "2":
+            jt: ["1", "3"]
+        "3":
+            jt: []
+        """
+        expected = """
+        "0":
+            jt: ["1"]
+        "1":
+            jt: ["2", "5"]
+        "2":
+            jt: ["6", "7"]
+        "3":
+            jt: []
+        "4":
+            jt: ["1", "3"]
+            be: ["1"]
+        "5":
+            jt: ["4"]
+        "6":
+            jt: ["4"]
+        "7":
+            jt: ["4"]
+        """
+        original_scfg, block_dict = SCFG.from_yaml(original)
+        expected_scfg, _ = SCFG.from_yaml(expected)
+        loop_restructure_helper(original_scfg, set({block_dict["1"], block_dict["2"]}))
+        self.assertSCFGEqual(expected_scfg, original_scfg)
+
     def test_double_exit(self):
         """Loop has two exiting blocks.
 

